apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js

```diff
@@ -1,41 +1,169 @@
+import { gradeAssessmentInstanceAsync } from '../../lib/assessment';
+import { setQuestionCopyTargets } from '../../lib/copy-question';
 import { getQuestionGroupPermissions } from '../../lib/groups';
+import { uploadFile, deleteFile } from '../../lib/file-store';
+import { idsEqual } from '../../lib/id';
+import { insertIssue } from '../../lib/issues';
+import { selectVariantById } from '../../models/variant';
 
-const util = require('util');
-const ERR = require('async-stacktrace');
-const _ = require('lodash');
-const express = require('express');
+const logPageView = promisify(LogPageView('studentInstanceQuestion'));
 const router = express.Router();
-const async = require('async');
-const asyncHandler = require('express-async-handler');
 
-const error = require('@prairielearn/error');
-const logPageView = require('../../middlewares/logPageView')('studentInstanceQuestion');
-const grading = require('../../lib/grading');
-const questionRender = require('../../lib/question-render');
-const assessment = require('../../lib/assessment');
-const studentInstanceQuestion = require('../shared/studentInstanceQuestion');
-const sqldb = require('@prairielearn/postgres');
-const { setQuestionCopyTargets } = require('../../lib/copy-question');
+/**
+ * Get a validated variant, or throw an exception.
+ *
+ * @param {string} unsafe_variant_id The unsafe variant ID to validate
+ * @param {string} instance_question_id The validated instance_question_id to validate against
+ * @returns {import('../../lib/db-types').Variant} The validated variant
+ */
+async function getValidVariant(unsafe_variant_id, instance_question_id) {
+  const variant = await selectVariantById(unsafe_variant_id);
+  if (!variant || !idsEqual(variant.instance_question_id, instance_question_id)) {
+    throw error.make(
+      400,
+      `Client-provided variant id "${unsafe_variant_id}" does not belong to the authorized instance question "${instance_question_id}"`,
+    );
+  }
+  return variant;
+}
+
+/**
+ * Get a validated variant ID from a request, or throw an exception.
+ *
+ * This function assumes req.body.__variant_id has been sent by the client, but
+ * it is currently untrusted. We check that it is a valid variant_id and
+ * belongs to the authorized res.locals.instance_question_id and return it if
+ * everything is ok. If anything is invalid or unauthorized, we throw an
+ * exception.
+ *
+ * @param {express.Request} req The request object
+ * @param {express.Response} res The response object
+ * @returns {string} The validated variant ID
+ */
+async function getValidVariantId(req, res) {
+  return (await getValidVariant(req.body.__variant_id, res.locals.instance_question.id)).id;
+}
+
+async function processFileUpload(req, res) {
+  if (!res.locals.assessment_instance.open) throw new Error(`Assessment is not open`);
+  if (!res.locals.authz_result.active) {
+    throw new Error(`This assessment is not accepting submissions at this time.`);
+  }
```

Comment:
Can we use non-500 status codes here? Ditto for these other `process...` functions.

Reply:
Done in 17500451b78b79451bb559cfa3b45c4952ce821a.